### DISCUSSION TITLE
Recommend currently available means for container HEALTHCHECK

### DISCUSSION THREAD

**Discussion by vit-zikmund** [keycloak/keycloak#36537] [Archive]
### Description

Hi, I was setting up keycloak in a docker compose deployment, where I wanted to have a working healthcheck. The [documentation says to install `curl`](https://www.keycloak.org/server/health#_healthcheck) in such cases, but then again, that's rather convoluted and [generally not recommended by another piece of docs](https://www.keycloak.org/server/containers#_installing_additional_rpm_packages). Compose doesn't have any external means of running a HTTP healthcheck, it merely offers to externally set the container's `HEALTHCHECK` directive, which runs a process on the same image.

Trying to follow the latter recommendation I managed to develop this still pretty small oneliner which does the job using only `bash`/`coreutils` means:

```sh
# the simplest variant using the old HTTP/1.0 protocol
{ printf 'HEAD /health/ready HTTP/1.0\r\n\r\n' >&0; grep 'HTTP/1.0 200'; } 0<>/dev/tcp/localhost/9000

# another variant using HTTP/1.1 protocol that that needs the TCP connection closed and requires a host header
{ printf 'HEAD /health/ready HTTP/1.1\r\nhost:\r\nconnection: close\r\n\r\n' >&0; grep 'HTTP/1.1 200'; } 0<>/dev/tcp/localhost/9000
```

I believe, when proved right, this could become the new recommendation in the docs, or potentially even a `HEALTHCKECK` directive in the default image.

### Motivation

To tighten any potential attack surface even for plain container deployments.

### Details

For the curious, here's how this works:
1. For the code within braces, a TCP connection is made to the keycloak's management port and a successful connection is redirected in the read-write fashion to the descriptor `0` (stdin).
   - When bash fails to connect (TCP RST), it ends up with an error right away
   - When the connection is hanging (no reply till TCP retry timeout, usually about 1 minute), it just hangs, virtually being a subject to the `HEALTHCHECK`'s timeout (which should be definitely smaller than the usual TCP retry timeout).
1. Then a simple hand-crafted HTTP `HEAD` request is sent to the socket using `printf`. This is supposed to always succeed, unless the send buffer of the socket is set ridiculously small on the target OS. In the other case it will just hang again, not being able to push all the bytes through, until that eventually happens or times out.
1. Next, the eventual response is being checked with `grep` to be the successful one. Only at this time it's return code (and the final) is 0.
   - When no response comes, it's hanging forever and is subject to timeout.
   - When a 503 response comes, `grep` doesn't match anything and returns 1.

In order not to have the "client" close the TCP connection (for HTTP/1.1) after a successfully read message (while still reading the full socket's read buffer), I send the `connection: close` header with the request, so the server closes the connection/socket once it fully responds. When this happens, `grep` ends on its own after getting an EOF from the socket.

Another option would be to call `grep -m1` to exit after the 1st match, but that would result in the socket being closed without its read buffer being fully read, which usually warrants a RST packet back to the server. I prefer the graceful  variant.

**Comment by col-panic** [comment]
I can't get this to work with 26.0.8? 

But there seems to be another problem too, as
```
bash-5.1$ exec 3<>/dev/tcp/127.0.0.1/9000;echo -e 'HEAD /health/ready HTTP/1.1\r\nhost: http://localhost\r\nConnection: close\r\n\r\n' >&3;cat <&3
HTTP/1.1 400 Bad Request
connection: close
content-length: 0
```

despite `KC_HEALTH_ENABLED=true`

**Comment by vit-zikmund** [comment]
Hi @col-panic, I have developed this for KC v25.something, but it should generally work in any newer version.
Maybe some security hardening was done in v26, so more request crafting should be done, but on the other hand it's just a health check, so I'm a bit unsure.

I see 2 differences in your example:
- You use the POSIXly badly defined `echo` instead of `printf`, well... knock yourself out :) but it adds one more `\n` in your output (unless you use the `-n` argument). This newline is definitely garbage and likely that's the reason for the 400.
- You add the `host` header. I think it's unnecessary, but if you insist, I believe it should be without the http:// prefix.

**Comment by col-panic** [comment]
@vit-zikmund thanks for your response! Actually I was doing this just to see what really is returned from `/health/ready` and even with your changes, which I guess amount to
```
bash-5.1$ exec 3<>/dev/tcp/127.0.0.1/9000;echo -en 'HEAD /health/ready HTTP/1.1\r\nConnection: close\r\n\r\n' >&3;cat <&3
HTTP/1.1 400 Bad Request
connection: close
```

it's not correct. And using your code does
```
bash-5.1$ { printf 'HEAD /health/ready HTTP/1.1\r\nconnection: close\r\n\r\n' >&0; grep 'HTTP/1.1 200'; } 0<>/dev/tcp/localhost/9000
bash-5.1$ echo $?
1
```

Could you check, if any of this works for you in Keycloak 26? I also added this as a discussion point in https://github.com/keycloak/keycloak/discussions/36526 - I just want to know if it's me or keycloak itself :)

**Reply by vit-zikmund** [reply]
Hi @col-panic I see you managed to figure this out real quick in your other discussion. For the local readers, the necessity of the `host` header for `HTTP/1.1` is the new requirement that got added. I've played with it a bit, and noticed that the value of the header is not checked to be in sync with the `--hostname` option. This check might get added in any future release, so one might make a little step back to the legacy `HTTP/1.0` protocol, which works well, too.
It's really the question what is more future-proof. Using a legacy HTTP protocol or an empty/bogus `host` header :) Ideally, the value should be the real hostname, but this puts only more burden on the administrator of such solution. Alas, it looks like we'll have to stay vigilant about what security hardening happens on this side. Till then, I'm leaning to use the `HTTP/1.0` protocol, which doesn't require any `host` header at all.

**Comment by shawkins** [comment]
Based upon the discussion with @col-panic moved this to a discussion as well.

**Reply by vit-zikmund** [reply]
Hello @shawkins, I've provided an updated command that works again (two variants actually). Is there some chance I could get the attention of someone who's responsible for the documentation, to review the proposal for the [HEALTHCHECK suggestion in the docs](https://www.keycloak.org/observability/health#_healthcheck)?

**Reply by shawkins** [reply]
@vit-zikmund you would open an issue which references this discussion, then if you are able open a pull request that references that issue.

**Reply by vit-zikmund** [reply]
[Done](https://github.com/keycloak/keycloak/issues/38126)

**Comment by darius-m** [comment]
According to https://github.com/quarkusio/quarkus/issues/37582#issuecomment-2462559462, the HTTP 1.1 protocol mandates the use of the `Host` header, which is missing from the excerpts. We were also seeing the "Unhandled exception in routing" in the Keycloak logs when we were not setting the header properly (we opted to use the public hostname of the SSO as the header's value in this case).
